ejb3/src/main/java/org/jboss/as/ejb3/timerservice/persistence/database/DatabaseTimerPersistence.java

```diff
@@ -844,15 +844,22 @@ public void run() {
                                             // remove and add -> the probable cause is db glitch
                                             EnumSet<TimerState> valid = EnumSet.of(TimerState.IN_TIMEOUT, TimerState.RETRY_TIMEOUT, TimerState.CREATED, TimerState.ACTIVE);
                                             boolean validDBTimer = valid.contains(holder.timer.getState());
-                                            boolean validMemoryTimer = !valid.contains(oldTimer.getState());
+                                            boolean validMemoryTimer = oldTimer != null && !valid.contains(oldTimer.getState());
                                             // if timers memory - db are in non intersect subsets of valid/invalid states. we put them in sync
                                             if (validMemoryTimer && validDBTimer) {
                                                 synchronized (DatabaseTimerPersistence.this) {
-                                                    Set<String> timers = knownTimerIds.get(timedObjectId);
-                                                    timers.remove(oldTimer.getId());
+                                                    knownTimerIds.get(timedObjectId).remove(oldTimer.getId());
                                                     listener.timerSync(oldTimer, holder.timer);
                                                 }
-
+                                            } else if (validDBTimer && oldTimer == null) {
+                                                // timer migration among nodes.
+                                                // persistence tier and timer subsystem is out of sync
+                                                // we were able to delete it from the persistence tier but it does not exists in timer service
+                                                // need to add it again to the persistence tier and register it in timer service if is valid
```

Comment:
I don't understand why "we were able to delete it from the persistence tier", because the `dbTimer` is just retrieved from db.

Also why the comment "need to add it again to the persistence tier"? It already exists in db and so we just need to add it to timer service of the current node, which is what the statement below does.

Reply:
@chengfang I ment this class. Sometimes the database timer persistence class is out of sync from the TimerService (it did have the timer associated but it was not registered in the timer service causing a null pointer exception when you are trying to get the timer service).. It is true that it might be missleading the comment